python/cuml/cuml/experimental/experimental/decomposition/kpca.pyx

```diff
@@ -0,0 +1,502 @@
+    eigenvectors_ = CumlArrayDescriptor(order='F')
+    trans_input_ = CumlArrayDescriptor(order='F')
+
+    @device_interop_preparation
+    def __init__(self, *, handle=None, n_components=None, kernel='linear', gamma=None,
+                degree=3, coef0=1, kernel_params=None, alpha=1.0,
+                fit_inverse_transform=False, eigen_solver='auto', tol=0,
+                max_iter=None, iterated_power=15, remove_zero_eig=False, n_jobs=None,
+                random_state=None, copy_X=True, verbose=False, output_type=None):
+        if fit_inverse_transform:
+            raise NotImplementedError("Inverse transform is not supported")
+        if random_state is not None:
+            raise NotImplementedError("Random state is not supported in available eigen solvers")
+        if n_jobs is not None and n_jobs != -1:
+            raise NotImplementedError("n_jobs does not apply to this algorithm")
+        if max_iter is not None:
+            raise NotImplementedError("max_iter is not supported in available eigen solvers. Use iterated_power for Jacobi solver")
+        super().__init__(handle=handle,
+                        verbose=verbose,
+                        output_type=output_type)
+        self.copy_X = copy_X
+        self.max_iter = max_iter
+        self.iterated_power = iterated_power
+        self.n_components_ = n_components
+        self.remove_zero_eig = remove_zero_eig
+        self.random_state = random_state
+        self.eigen_solver = eigen_solver
+        self.tol = tol
+        self.kernel = kernel
+        self.c_kernel = self._get_c_kernel(kernel)
+        self.c_algorithm = self._get_algorithm_c_name(self.eigen_solver)
+        self.gamma_ = gamma
+        self.degree = degree
+        self.coef0 = coef0
+        self.alpha = alpha
+        self.fit_inverse_transform = fit_inverse_transform
+
+        self.trans_input_ = None
+        self.eigenvectors_ = None
+        self.eigenvalues_ = None
+
+    def _get_c_kernel(self, kernel):
+        """
+        Get KernelType from the kernel string.
+
+        Parameters
+        ----------
+        kernel: string, ('linear', 'poly', 'rbf', or 'sigmoid')
+        """
+        return {
+            'linear': LINEAR,
+            'poly': POLYNOMIAL,
+            'rbf': RBF,
+            'sigmoid': TANH
+        }[kernel]
+
+    def _get_algorithm_c_name(self, algorithm):
+        IF GPUBUILD == 1:
+            algo_map = {
+                'full': Solver.COV_EIG_DQ,
+                'auto': Solver.COV_EIG_DQ,
+                'jacobi': Solver.COV_EIG_JACOBI
+            }
+            if algorithm not in algo_map:
+                msg = "algorithm {!r} is not supported"
+                raise TypeError(msg.format(algorithm))
+
+            return algo_map[algorithm]
+
+    def _build_params(self, n_rows, n_cols):
+        IF GPUBUILD == 1:
+            cdef paramsKPCA *params = new paramsKPCA()
+            params.n_components = min(self.n_components_ or n_rows, n_rows)
```

Comment:
[Sklearn reference](https://github.com/scikit-learn/scikit-learn/blob/b72e81af473c079ae95314efbca86557a836defa/sklearn/decomposition/_kernel_pca.py#L333-L336) for setting n_components